### DISCUSSION TITLE
blockcheck via libcurl build with custom cfilter

### DISCUSSION THREAD

**Discussion by EnergoStalin** [bol-van/zapret#704] [Ideas]
В теории помощью connection filters в libcurl возможно кросплатформенно распаралелить тестинг вставив условный cf-zapret между условными [cf-ssl и cf-socket](https://github.com/curl/curl/blob/46bd595b772f8f6856c2beb6ecc4dca290d53f4b/docs/internals/CONNECTION-FILTERS.md?plain=1#L34) где уже будет находится буффер готовый для модификации на лету без таблиц маршрутизаций и nfq в частности.
 
 Или проще говоря модифицировать то что сам курл отправляет в кастомном коннекшен фильтре.

 Сразу в голову приходит что скорее всего всякие ipfrag работать не будут хотя тоже спорно смотря на сколько глубоко внедриться фильтры позволяют. В ближайшие пару дней попробую копнуть глубже и выяснить возможно ли приложив адекватные усилия пропатчить libcurl сохранив вменяемую поддрежку протоколов так что бы можно было прям в нём модифицировать tcp/udp пакеты на достаточно низком уровне.

А пока хотелось бы услышать мнение более сведующих в теме людей.

### Для референса cfilter
- [api](https://github.com/curl/curl/blob/78c3172921f6e42143638a894408b729e8038f32/lib/cfilters.h)
- [usage](https://github.com/search?q=repo%3Acurl%2Fcurl+cfilter+NOT+path%3A*.h&type=code)
- [docs](https://github.com/curl/curl/blob/46bd595b772f8f6856c2beb6ecc4dca290d53f4b/docs/internals/CONNECTION-FILTERS.md?plain=1)

**Comment by HentayHentay** [comment]
libcurl станет узким местом. При обновлении curl или libcurl продеться каждый раз завозить патчи 
 
Да и в силу многообразия операционных систем, может отсутствовать та или иная возможность к применению патча

**Reply by EnergoStalin** [reply]
С отлаженным CI и минималистичными патчами обновится не должно быть проблемой. Vencord же как то живёт. При должном наличии энузиазма на пару дней максимум релиз опоздает что не так критично для узконаправленной утилиты тем более что cfilter api выглядит стабильным и широко используется в самом libcurl.

**Comment by bol-van** [comment]
С блокчеком планируется распараллеливание теста множества доменов на 1 стратегии.
Это не так сложно сделать.

Усложнять блокчек какими-то сложными зависимостями не планируется

Текущий блокчек основывается на действиях реальных tpws и nfqws, а не на собственном конструировании измененных запросов. Такой подход не покроет и малой доли всех методов дурения, реализуемых zapret

**Reply by EnergoStalin** [reply]
Пока что мне кажется что при условии raw траффика не отбрасываемого ос и доступа к сокету на уровне голых протоколов 3 уровня(возможность получить байтики сформированного udp/tcp пакета) должно сработать везде. В теории конечно на винде драйвер находится ниже файервола который может дропать пакеты в отличие от программы работающей в юзермоде но это тоже в теории должно быть решаемо добавлением в исключения. Меня больше беспокоит именно объем работ и реализуемость поставленных выше условий средствами libcurl.

Поскольку cfilter позволяет так же как и условный nfq дропать/модифицировать пакеты. Я считаю что proof of concept попробовать сделать как минимум стоит.

Это может быть не так удобно с точки зрения дебаггинга приложения целиком что тоже спорно поскольку для стратегий подходит идеально а cli часть с фильтром можно и юнит тестами погонять. Опять же с перспективы конечного пользователя этого должно быть достаточно для выявления стратегий практически мгновенно по сравнению с текущим подходом если оно работает так как я описал выше что ещё предстоит выяснить.

Я бы и не подумал в это лезть если бы блокчек занимал адекватное количество времени. Распараллеливание по доменам это хорошо от него я тоже не отказываюсь. Но я конкретно загорелся этой идеей из за твоего [коммента](https://github.com/bol-van/zapret/issues/647#issuecomment-2439549592).

**Reply by EnergoStalin** [reply]
Это конечно чисто построение воздушных замков на горах воды пока не будет работающего proof of concept.

Не говоря о том что я просто делюсь идеей и не перекладываю ни на кого ни её поддержку ни реализацию.

Я вижу новый флоу блокчекинга отдельной надстройкой над libcurl с C API для возможных биндингов на языки с более богатой стандартной библиотекой по типу Rust. С несколькими версиями более совместимые для роуторов и более красивые для ПК.

**Reply by bol-van** [reply]
Я не представляю как возможно смешивать работу на уровне сокетов и низкоуровневое ковыряние в L3/L4.
Последним занимается ОС, а не libcurl

**Reply by EnergoStalin** [reply]
Если я правильно понимаю zapret не спускается на L2 уровень а это значит что надо только завести готовую реализацию UDP/TCP протокола поверх libcurl оборачивая соединение по RAW сокету вручную в него перед отправкой отдавая буффер на дурение.

На винде однако придётся всё же [обернуть](https://learn.microsoft.com/en-us/windows/win32/winsock/tcp-ip-raw-sockets-2?redirectedfrom=MSDN#limitations-on-raw-sockets) ещё во что то что бы пакет не считался TCP разворачивая его в windrivert коллбэке.

На мак и bsd я даже не смотрел пока но они же unix right?

Кстати пока читал мне в голову пришла более здравая мысль. Раскидать по виртуальным интерфейсам вызывая курл с `--interface` флагом накинув уже на них фильтры. Где на 1 интерфейс 1 инстанс запрета с разными флагами. Как раз windrivert слушает на уровне [network](https://github.com/bol-van/zapret/blob/4374ca229084c330d5c85ec1b3a6596217b66a9c/nfq/darkmagic.c#L1349) который поддерживается [`ifIdx`](https://reqrypt.org/windivert-doc.html#filter_language) фильтром. Так что я бы взял этот вариант на вооружение.

**Reply by bol-van** [reply]
> завести готовую реализацию UDP/TCP протокола

А может весь linux перепишем заново под libcurl? Круто же будет выглядить. curlix
Это монстрообразная задача, чудовищная по обьему усилий. UDP еще довольно несложно, но вот tcp переписать - дело тяжелое.

Вы вообще не в ту сторону смотрите. Никто так не делает. Зачем переделывать то, что уже десятками лет отточено ?
Я же описал возможный вариант распараллеливания. Раскидать их по UID, и на каждый UID свое правило tables, свой инстанс nfqws. Это работает на всех *nix, но не на винде

**Reply by EnergoStalin** [reply]
# ZaAstraCurl Linux
> А может весь linux перепишем заново под libcurl? Круто же будет выглядить.

Ну тут уделал. Жаль такими амбициями не обладаю.

Если уж и пытатся как то продолжать плясать через RAW сокеты без своего TCP/UDP то придётся свой протокол городить на RAW сокетах который спрячет в себя реальный пункт назначения пакета и будет выступать как прокси на RAW сокетах. пересобирая его по кусочкам с заменой destination ip хедера. Но я так делать не планирую.

# Windows
> но не на винде

Мне казалось что мы обозначили что проблема только в винде. Перечитал понял что нет. Обозначаю "проблема только в винде".

Под Unix оно и удобно и понятно там уйма способов. За поддержку виртуальных интерфейсов на всех *nix конечно не ручаюсь этим тоже драйвер заниматься должен которого может и не быть. Но если выбирать унифицированный подход то этот должно работать везде при наличии драйвера.

Для винды подход через создание виртуальных интерфейсов похоже что единственный и самый логичный. Сразу бросается в глаза вариант с HyperV но он не у всех есть и только от Home Pro тира на сколько я помню его можно включить. Так что я бы ещё посмотрел варианты.

В случае с интерфейсным распаралеливанием должно хватить windrivert [ifIdx](https://reqrypt.org/windivert-doc.html#filter_language) фильтра. И `--interface` флага curl.

# Поддержка нескольких ОС
Вообще у меня есть сомнения на счёт необходимости кросплатформенно тестить поскольку единственное что будет отличатся от системы к системе это реализация L2-3 нетворк стека ну и немного TCP пейлоад который в большинстве случаев зашифрован TLS и не проверяется. Так что может есть смысл сфокусироватся на удобном решении для *nix в первую очередь не смотря на винду.

Мне эксклюзивное решение под линукс подойдёт. Я думал ты хочешь себе поддержку распаралеливания стратегий на винде. Неправильно тебя понял?

**Reply by bol-van** [reply]
Я в эти дебри, описанные выше, не полезу.  Это гиблое дело.
Чудовищные усилия при непонятном результате.
Мне проще сказать ставьте виртуалку с linux, если хотите побыстрее

**Reply by EnergoStalin** [reply]
Для винды и так по сути только goodcheck существует в качестве тестера не считая bat скриптов. Который я очень хочу завезти на линукс.

Если резюмировать всё что потребуется непосредственно от winws то это будет "опция установки windrivert фильтра по указанному интерфейсу". Остальное прикручивается тривиально за исключением самого создания интерфейса что я вынес бы за рамки этого репозитория.

Как то неловко на самом деле получилось в итоге всё сошлось на там на чём и предыдущий тред. Нужна отдельная быстрая утилита для пользовательских тестов.
